### PR TITLE
HIVE-26825 : Compactor: Cleaner shouldn't fetch table details again a…

### PR DESCRIPTION
**Pull request by rkirtir**
…nd again for partitioned tables

### What changes were proposed in this pull request?
(https://issues.apache.org/jira/browse/HIVE-26825)

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
junit

### CODE REVIEW

**Review by veghlaci05 — CHANGES_REQUESTED** [review]

File: ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java [hunk 1]
```diff
@@ -1093,5 +1095,34 @@ public void testReady() throws Exception {
...
+    doThrow(new RuntimeException()).when(mockedHandler).markCleaned(nullable(CompactionInfo.class));
+    Cleaner cleaner = Mockito.spy(new Cleaner());
+    cleaner.setConf(conf);
+    cleaner.init(new AtomicBoolean(true));
+    cleaner.run();
+
+    ShowCompactResponse rsp = txnHandler.showCompact(new ShowCompactRequest());
+    List<ShowCompactResponseElement> compacts = rsp.getCompacts();
+    Assert.assertEquals(1, compacts.size());
+    Mockito.verify(cleaner, times(1)).resolveTable(Mockito.any());
```
[the remark this answers is not in the document]
Comment (rkirtir) [reply ql/src/test/org/apache/hadoop/hive/ql/txn/compactor/TestCleaner.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java [hunk 2]
```diff
@@ -96,6 +100,7 @@ public class Cleaner extends MetaStoreCompactorThread {
   static final private String CLASS_NAME = Cleaner.class.getName();
   static final private Logger LOG = LoggerFactory.getLogger(CLASS_NAME);
   private boolean metricsEnabled = false;
+  private Optional<Cache<String, TBase>> metaCache = Optional.empty();
```
Comment (veghlaci05) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:103]: You could use an abstract boolean useCache() method in MetastoreCompactorThread, (and implement in Initiator and Cleaner respectively) and move the metaCache field and the init logic there. As a result computeIfAbsent should not take metaCache as an argument, and caching could be hidden from the subclasses.

**Review by veghlaci05 — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java (hunk 2 above)
Comment (rkirtir) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Cleaner.java:99]: I have added 
abstract Optional<Cache<String, TBase>> initializeCache(boolean tableCacheOn); in  MetastoreCompactorThread(Had to made MetastoreCompactorThread abstract class now)

Initiator and Cleaner now implements initializeCache based on tableCacheOn property

**Review by veghlaci05 — APPROVED** [review]
LGTM, only one minor thing.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MetaStoreCompactorThread.java [hunk 3]
```diff
@@ -133,4 +141,27 @@ protected static long updateCycleDurationMetric(String metric, long startedAt) {
...
+  }
+
+  Optional<Cache<String, TBase>> initializeCache(boolean tableCacheOn) {
+    if (tableCacheOn) {
+      metaCache = Optional.of(CacheBuilder.newBuilder().softValues().build());
+    }
+    return metaCache;
+  }
+
+  protected void invalidateMetaCache(){
```
[the remark this answers is not in the document]
Comment (rkirtir) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/MetaStoreCompactorThread.java]: done

**Pull request merged**